lib/chef/knife/cloud/google_service.rb

```diff
@@ -0,0 +1,491 @@
+      check_api_call { connection.get_machine_type(project, zone, machine_type) }
+    end
+
+    def valid_network?(network)
+      return false if network.nil?
+      check_api_call { connection.get_network(project, network) }
+    end
+
+    def image_exist?(image_project, image_name)
+      check_api_call { connection.get_image(image_project, image_name) }
+    end
+
+    def valid_public_ip_setting?(public_ip)
+      public_ip.downcase! if public_ip.respond_to?(:downcase)
+
+      if public_ip.nil? || public_ip == "ephemeral" || public_ip == "none"
+        true
+      elsif valid_ip_address?(public_ip)
+        true
+      else
+        false
+      end
+    end
+
+    def valid_ip_address?(ip_address)
+      IPAddr.new(ip_address)
+    rescue IPAddr::InvalidAddressError
+      false
+    else
+      true
+    end
+
+    def instance_object_for(options)
+      inst_obj                    = Google::Apis::ComputeV1::Instance.new
+      inst_obj.name               = options[:name]
+      inst_obj.can_ip_forward     = options[:can_ip_forward]
+      inst_obj.disks              = instance_disks_for(options)
+      inst_obj.machine_type       = machine_type_url_for(options[:machine_type])
+      inst_obj.metadata           = instance_metadata_for(options[:metadata])
+      inst_obj.network_interfaces = instance_network_interfaces_for(options)
+      inst_obj.scheduling         = instance_scheduling_for(options)
+      inst_obj.service_accounts   = instance_service_accounts_for(options) unless instance_service_accounts_for(options).nil?
+      inst_obj.tags               = instance_tags_for(options[:tags])
+
+      inst_obj
+    end
+
+    def instance_disks_for(options)
+      disks = []
+      disks << instance_boot_disk_for(options)
+      options[:additional_disks].each do |disk_name|
+        begin
+          disk = connection.get_disk(project, zone, disk_name)
+        rescue Google::Apis::ClientError => e
+          ui.error("Unable to attach disk #{disk_name} to the instance: #{e.message}")
+          raise
+        end
+
+        disks << Google::Apis::ComputeV1::AttachedDisk.new.tap { |x| x.source = disk.self_link }
+      end
+
+      disks
+    end
+
+    def instance_boot_disk_for(options)
+      disk = Google::Apis::ComputeV1::AttachedDisk.new
+      params = Google::Apis::ComputeV1::AttachedDiskInitializeParams.new
+
+      disk.boot           = true
+      disk.auto_delete    = options[:boot_disk_autodelete]
+      params.disk_name    = boot_disk_name_for(options)
+      params.disk_size_gb = options[:boot_disk_size]
+      params.disk_type    = disk_type_url_for(boot_disk_type_for(options))
+      params.source_image = image_search_for(options[:image], options[:image_project])
+
+      disk.initialize_params = params
+      disk
+    end
+
+    def boot_disk_type_for(options)
+      options[:boot_disk_ssd] ? "pd-ssd" : "pd-standard"
+    end
+
+    def image_search_for(image, image_project)
```

Comment:
It would be nice if GCE provided an API to fetch the list of official project names for images. I've seen the practice of keeping a list of known projects in a hash in other project such as libcloud,

e.g. https://github.com/apache/libcloud/blob/trunk/libcloud/compute/drivers/gce.py#L1016

With libcloud, users can also specify an image as 'centos' or 'redhat' to pick up the most current image without needing to know the exact name.

Definitely not required, just pointing out other project's workaround with GCE.